gneiss/_formula.py

```diff
@@ -0,0 +1,90 @@
+# ----------------------------------------------------------------------------
+# Copyright (c) 2016--, gneiss development team.
+#
+# Distributed under the terms of the GPLv3 License.
+#
+# The full license is in the file COPYING.txt, distributed with this software.
+# ----------------------------------------------------------------------------
+import pandas as pd
+import statsmodels.formula.api as smf
+from skbio.stats.composition import ilr
+from gneiss.util import match, match_tips, rename_internal_nodes
+from gneiss._summary import RegressionResults
+from gneiss.balances import balance_basis
+
+
+def _process(table, metadata, tree):
+    """ Matches tips, features and samples between the table, metadata
+    and tree.  This module returns the features and samples that are
+    contained in all 3 objects.
+    """
+    _table, _metadata = match(table, metadata)
```

Comment:
Thanks for the explanation, I agree that avoiding side-effects is a good idea. As for avoiding making copies, I was referring to not needing to create `_table` and `_metadata` and instead just referring to `table` by overwriting the variable in return i.e. `table, metadata = match(table, metadata)` instead of `_table, _metadata = match(table, metadata)`. No side-effects and no duplication of variables in the scope of that function.

instead of: